Sources/SwiftFormatPrettyPrint/PrettyPrint.swift

```diff
@@ -43,66 +31,129 @@ public class PrettyPrinter {
-  private var currentGroupColorString: String {
-    return Ansi.color(currentDebugGroupMarkerColor + 1)
+  /// Keeps track of the most recent number of consecutive newlines that have been printed.
+  ///
+  /// This value is reset to zero whenever non-newline content is printed.
+  private var consecutiveNewlineCount = 0
+
+  /// Keeps track of the most recent number of spaces that should be printed before the next text
+  /// token.
+  private var pendingSpaces = 0
+
+  /// Indicates whether or not the printer is currently at the beginning of a line.
+  private var isAtStartOfLine = true
+
+  /// The kind of the last break token that was printed.
+  private var lastBreakKind: BreakKind = .reset
+
+  /// The computed indentation level, as a number of spaces, based on the state of any unclosed
+  /// delimiters and whether or not the current line is a continuation line.
+  private var currentIndentation: [Indent] {
+    var totalIndentation = indentationStack
+    if currentLineIsContinuation {
+      totalIndentation.append(configuration.indentation)
+    }
+    return totalIndentation
   }
 
   /// Creates a new PrettyPrinter with the provided formatting configuration.
   ///
   /// - Parameters:
   ///   - context: The formatter context.
   ///   - node: The node to be pretty printed.
-  public init(context: Context, node: Syntax, isDebugMode: Bool, printTokenStream: Bool) {
+  public init(context: Context, node: Syntax, printTokenStream: Bool) {
     self.context = context
     let configuration = context.configuration
     self.tokens = node.makeTokenStream(configuration: configuration)
     self.maxLineLength = configuration.lineLength
-    self.isDebugMode = isDebugMode
     self.spaceRemaining = self.maxLineLength
     self.printTokenStream = printTokenStream
   }
 
-  /// Append the input string to the output buffer
-  func write<S: StringProtocol>(_ str: S) {
+  /// Append the given string to the output buffer.
+  ///
+  /// No further processing is performed on the string.
+  private func writeRaw<S: StringProtocol>(_ str: S) {
     outputBuffer.append(String(str))
   }
 
+  /// Writes the given number of newlines to the output stream.
+  ///
+  /// - Parameters:
+  ///   - count: The number of newlines to write.
+  ///   - discretionary: Indicates whether the newlines are user-entered discretionary newlines.
+  ///     Discretionary newlines are always printed (up through the maximum allowed number provided
+  ///     to the printer at initialization time). Non-discretionary newlines are only printed if
+  ///     discretionary newlines have already not been printed yet.
+  private func writeNewlines(_ count: Int, discretionary: Bool) {
+    // We add 1 because it takes 2 newlines to create a blank line.
+    let maximumNewlines = configuration.maximumBlankLines + 1
+    let numberToPrint: Int
+    if consecutiveNewlineCount + count <= maximumNewlines {
+      numberToPrint = count - consecutiveNewlineCount
```

Comment:
If `consecutiveNewlineCount` is non-zero (meaning we've already written some newlines, perhaps ones inserted by the token stream creator as opposed to discretionary ones), then we need to subtract that off from the next time we print so that we don't exceed the maximum number allowed by the configuration.

Reply:
Don't we already guarantee that we don't exceed the maximum because of the `if` statement?

Reply:
I took a closer look at this and it turns out it's the condition in the `if` statement that I had wrong. The new commit fixes it and updates the comment to try to provide a better description of the intent here. The general idea is that you might have a break that fires (inserting a newline) immediately before a `newlines` token that represents discretionary line breaks, but the break fired _because_ of those discretionary line breaks contributing to the break's length. In that case, we get one of the newlines we want from the break, and then we only need to print one fewer of the discretionary line breaks to get the desired total.

In that case, the condition I want is `count <= maximumNewLines` because `count` represents the total number of desired consecutive newlines, and I still need to subtract off the number we have already seen so far.